target_chains/stylus/contracts/pyth-receiver/src/structs.rs

```diff
@@ -1,14 +1,72 @@
 use alloc::vec::Vec;
-use stylus_sdk::{prelude::*, storage::{StorageU16, StorageU64, StorageI32, StorageI64, StorageFixedBytes}};
-use stylus_sdk::alloy_primitives::{U64, I32, I64};
+use stylus_sdk::{
+    prelude::*,
+    storage::{
+        StorageU64, StorageI32, StorageI64, StorageU16, StorageFixedBytes, StorageKey
+    },
+};
+use stylus_sdk::alloy_primitives::{U16, FixedBytes,U64, I32, I64, B256, U256, keccak256};
+use pythnet_sdk::messages::PriceFeedMessage;
 
-// DataSource struct to store chain/emitter pairs
+#[derive(Debug)]
 #[storage]
 pub struct DataSourceStorage {
     pub chain_id: StorageU16,
     pub emitter_address: StorageFixedBytes<32>,
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct DataSource {
+    pub chain_id: U16,
+    pub emitter_address: FixedBytes<32>,
+}
+
+impl StorageKey for DataSourceStorage {
+    fn to_slot(&self, root: B256) -> U256 {
+        let mut bytes = [0u8; 34];
+
+        let chain_id: u16 = self.chain_id.get().to::<u16>();
+        // now you can use `chain_id` as a regular u16
+        let chain_id_bytes = chain_id.to_be_bytes();
+        
+        bytes[0..2].copy_from_slice(&chain_id_bytes);
+        bytes[2..].copy_from_slice(self.emitter_address.get().as_slice());
```

Comment:
an alternative is to use our serde to do this conversion instead of manual rolling and unrolling